### PR TITLE
Add SpecialMigration model (#7054 pt. 1)

### PR DESCRIPTION
**Pull request by yakkomajuri**
## Changes

I've decided to split up #7364's current and future changes, otherwise I can see it dragging.

Hence, here's a first PR we can merge with the DB model for special migrations. The one review aspect I'd like here is just: are we missing a column? What else would be useful?

### CODE REVIEW

File: latest_migrations.manifest [hunk 1]
```diff
@@ -3,7 +3,7 @@ auth: 0012_alter_user_first_name_max_length
 axes: 0006_remove_accesslog_trusted
 contenttypes: 0002_remove_content_type_name
 ee: 0005_project_based_permissioning
-posthog: 0184_delete_sessionsfilter
+posthog: 0185_auto_20211130_1055
```
Comment (yakkomajuri) [inline latest_migrations.manifest:6]:
```suggestion
posthog: 0185_special_migrations
```

File: posthog/models/special_migration.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    Errored = 3
+    RolledBack = 4
+
+
+class SpecialMigration(models.Model):
+    class Meta:
+        constraints = [models.UniqueConstraint(fields=["name"], name="unique name",)]
+
+    id: models.BigAutoField = models.BigAutoField(primary_key=True)
+    name: models.CharField = models.CharField(max_length=400, null=False, blank=False)
```
Comment (guidoiaquinti) [inline posthog/models/special_migration.py:18]: Nit: I know we don't expect this table to be huge in length but 400 looks a bit a waste to me. We could add an additional column `description` if we want to add more information to it

File: posthog/models/special_migration.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    NotStarted = 0
+    Running = 1
+    CompletedSuccessfully = 2
+    Errored = 3
+    RolledBack = 4
+
+
+class SpecialMigration(models.Model):
+    class Meta:
+        constraints = [models.UniqueConstraint(fields=["name"], name="unique name",)]
```
Comment (guidoiaquinti) [inline posthog/models/special_migration.py:15]: Do we need this?

File: posthog/models/special_migration.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+    id: models.BigAutoField = models.BigAutoField(primary_key=True)
+    name: models.CharField = models.CharField(max_length=400, null=False, blank=False)
+    progress: models.IntegerField = models.IntegerField(null=False, blank=False, default=0)
```
Comment (guidoiaquinti) [inline posthog/models/special_migration.py:19]: Nit: do we need this? Isn't this `progress == !status`?

File: posthog/models/special_migration.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    status: models.PositiveSmallIntegerField = models.PositiveSmallIntegerField(
+        null=False, blank=False, default=MigrationStatus.NotStarted
+    )
+
+    current_operation_index: models.PositiveSmallIntegerField = models.PositiveSmallIntegerField(
```
Comment (guidoiaquinti) [inline posthog/models/special_migration.py:24]: Ditto

File: posthog/models/special_migration.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    )
+    current_query_id: models.CharField = models.CharField(max_length=400, null=False, blank=False, default="")
+    celery_task_id: models.CharField = models.CharField(max_length=400, null=False, blank=False, default="")
+
+    started_at: models.DateTimeField = models.DateTimeField(null=True, blank=True)
+
+    # Can finish with status 'CompletedSuccessfully', 'Errored', or 'RolledBack'
+    finished_at: models.DateTimeField = models.DateTimeField(null=True, blank=True)
+
+    error: models.TextField = models.TextField(null=True, blank=True)
```
Comment (guidoiaquinti) [inline posthog/models/special_migration.py:35]: this should be `last_error` if we are thinking to update it without appending

File: posthog/models/special_migration.py (hunk 3 above)
Comment (yakkomajuri) [reply posthog/models/special_migration.py:15]: I believe it's good to have yes

File: posthog/models/special_migration.py (hunk 4 above)
Comment (yakkomajuri) [reply posthog/models/special_migration.py:19]: Progress is out of 100. It's for running migrations, so we know how far along they are. Migration definitions provide a function to check progress. e.g. 10000/25000 events moved to the new table

File: posthog/models/special_migration.py (hunk 5 above)
Comment (yakkomajuri) [reply posthog/models/special_migration.py:24]: This is very important. It helps us keep track of what op we're at. 

TL;DR:

- **Status:** one of NotStarted, Running, CompletedSuccessfully, Errored, RolledBack (and maybe more to come)
- **Progress:** Integer between 0 and 100 indicating the true progress of a migration 
- **Current op index:** _n_ where `n >=0 and n <= len(operations)`. This is _not_ real progress. Think about the example migration in the other PR. The first ops should complete in milliseconds, and so should the last few. There's one op in the middle that can take up days. This is the one column that isn't really "meta" - we actually depend on it to know what to run next, show the user where we failed, resume operations on crashes, etc.

File: posthog/models/special_migration.py (hunk 6 above)
Comment (yakkomajuri) [reply posthog/models/special_migration.py:35]: Well at least the way I'm building this, there will only ever be one error to deal with.

Migration starts, if it errors, we bail and show the error. If we restart, this is cleared. Indeed, I can see how it is the last error but given that you only get a new error if you re-run, the distinction seems clear to me whereas `last_error` might actually be confusing?

File: posthog/models/special_migration.py (hunk 2 above)
Comment (yakkomajuri) [reply posthog/models/special_migration.py:18]: Description is a good idea indeed. Will be very useful in the UI.

File: posthog/models/special_migration.py (hunk 3 above)
Comment (guidoiaquinti) [reply posthog/models/special_migration.py:15]: Can you explain why? Why do we need this additional constraint? Isn't the PK on `id` already enough?

File: posthog/models/special_migration.py (hunk 5 above)
Comment (guidoiaquinti) [reply posthog/models/special_migration.py:24]: Ah I thought we wanted to keep the migrations sequential and with max concurrency hardcoded to 1. In this case it make sense I think

File: posthog/models/special_migration.py (hunk 6 above)
Comment (guidoiaquinti) [reply posthog/models/special_migration.py:35]: Sure but I think what you described is an implementation detail: when the migration starts you already update a bunch of columns, you could technically blank this as well to archive the consistency you were mentioning.

File: posthog/models/special_migration.py (hunk 3 above)
Comment (yakkomajuri) [reply posthog/models/special_migration.py:15]: Essentially, I want to prevent at the db-level the case where we have 2 or more instances of a migration in the database, which can be at different statuses etc. 

We can ofc guard against this in code, but I think it's worth having the constraint. Although as I'm typing Im thinking we could get rid of ID altogether and make the name the PK?

Comment (guidoiaquinti) [reply posthog/models/special_migration.py:15]: 👍 ok if we think there's a possibility we might end up with duplicate migrations

File: posthog/models/special_migration.py (hunk 6 above)
Comment (yakkomajuri) [reply posthog/models/special_migration.py]: Updated to `last_error`

File: posthog/models/special_migration.py (hunk 2 above)
Comment (yakkomajuri) [reply posthog/models/special_migration.py]: Added description, limited to 50 + a description with 400 max

**Comment by yakkomajuri** [comment]
all addressed @guidoiaquinti, thanks for the review

File: posthog/migrations/0185_special_migrations.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    ]
+
+    operations = [
+        migrations.CreateModel(
+            name="SpecialMigration",
+            fields=[
+                ("id", models.BigAutoField(primary_key=True, serialize=False)),
+                ("name", models.CharField(max_length=50)),
+                ("description", models.CharField(blank=True, max_length=400, null=True)),
+                ("progress", models.IntegerField(default=0)),
```
Comment (tiina303) [inline posthog/migrations/0185_special_migrations.py:19]: why isn't this `PositiveSmallIntegerField` too?

**Review by tiina303 — APPROVED** [review]

File: posthog/migrations/0185_special_migrations.py (hunk 7 above)
Comment (yakkomajuri) [reply posthog/migrations/0185_special_migrations.py]: good catch, updated
